src/main/java/seedu/address/logic/commands/UndoCommand.java

```diff
@@ -26,82 +23,92 @@ public class UndoCommand extends Command {
             + "Example: " + COMMAND_WORD;
 
     public static final String MESSAGE_UNDO_COMMAND_SUCCESS = "Undo successful:\n%s";
-    public static final String MESSAGE_UNDO_ADD = "%s has been unadded from SocialBook";
+    public static final String MESSAGE_UNDO_COMMAND_NEUTRAL = "No action to undo:\n%s";
+    public static final String MESSAGE_UNDO_ADD = "%s has been removed from SocialBook";
     public static final String MESSAGE_UNDO_EDIT = "Edits to %s has been reverted";
-    public static final String MESSAGE_UNDO_DELETE = "%s have been added back to SocialBook";
+    public static final String MESSAGE_UNDO_DELETE =
+            "%s and their appointments have been added back to SocialBook";
     public static final String MESSAGE_UNDO_CLEAR = "Here is the list before clearing";
-    private final ArrayList<Command> pastCommands;
+    public static final String MESSAGE_UNDO_DELETE_APPOINTMENT =
+            "Appointment for %s has been added back to appointments";
+    public static final String MESSAGE_UNDO_ADD_APPOINTMENT =
+            "Appointment for %s has been removed from appointments";
+    private final CommandHistory pastCommands;
     private final Logger logger = LogsCenter.getLogger(LogicManager.class);
 
     /**
      * Loads the past commands into constructor.
      *
      * @param pastCommands all the past commands during this code run.
      */
-    public UndoCommand(ArrayList<Command> pastCommands) {
+    public UndoCommand(CommandHistory pastCommands) {
         this.pastCommands = pastCommands;
 
     }
 
     @Override
     public CommandResult execute(Model model) throws CommandException {
         requireNonNull(model);
-        if (pastCommands.isEmpty()) {
+        if (pastCommands.getCommandHistory().isEmpty()) {
             throw new CommandException(Messages.MESSAGE_NO_LATEST_COMMAND);
         }
-        Command latestCommand = pastCommands.get(pastCommands.size() - 1);
+        Command latestCommand = pastCommands.getCommandHistory().get(pastCommands.getSize() - 1);
         String latestCommandWord = latestCommand.getCommandWord();
         logger.info("----------------[COMMAND UNDONE][" + latestCommandWord + "]");
         String resultMessage = String.format(
                 "No change as command undone (%s) was not an action command", latestCommandWord);
+        boolean isAction = false;
 
         switch (latestCommandWord) {
         case "add":
             AddCommand addCommand = (AddCommand) latestCommand;
```

Comment:
Is the casting necessary here? `latestCommand.undo(model, pastCommands)` can be called directly due to polymorphism.